### PR TITLE
[Xtensa] Lowering FRAMEADDR/RETURNADDR operations.

### PR DESCRIPTION
**Pull request by andreisfr**

### CODE REVIEW

**Comment by andreisfr** [comment]
@arsenm , @s-barannikov , PTAL to this PR whenever you have the time.

File: llvm/lib/Target/Xtensa/XtensaISelLowering.cpp [hunk 1]
```diff
@@ -594,6 +594,26 @@ SDValue XtensaTargetLowering::LowerSELECT_CC(SDValue Op,
...
+  // check the depth
+  // TODO: xtensa-gcc can handle this, by navigating through the stack, we
+  // should be able to do this too
+  assert((cast<ConstantSDNode>(Op.getOperand(0))->getZExtValue() == 0) &&
+         "Return address can be determined only for current frame.");
+
+  MachineFunction &MF = DAG.getMachineFunction();
+  MachineFrameInfo &MFI = MF.getFrameInfo();
+  EVT VT = Op.getValueType();
+  unsigned RA = Xtensa::A0;
```
Comment (arsenm) [inline llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:608]: No reason to use this variable (but it should be Register/MCRegister)

File: llvm/lib/Target/Xtensa/XtensaISelLowering.cpp [hunk 2]
```diff
@@ -594,6 +594,26 @@ SDValue XtensaTargetLowering::LowerSELECT_CC(SDValue Op,
...
+  MFI.setReturnAddressIsTaken(true);
+
+  // Return RA, which contains the return address. Mark it an implicit
+  // live-in.
+  unsigned Register = MF.addLiveIn(RA, getRegClassFor(MVT::i32));
```
Comment (arsenm) [inline llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:613]: Type is Register

File: llvm/lib/Target/Xtensa/XtensaISelLowering.cpp [hunk 3]
```diff
@@ -594,6 +594,26 @@ SDValue XtensaTargetLowering::LowerSELECT_CC(SDValue Op,
                      FalseValue, TargetCC);
 }
 
+SDValue XtensaTargetLowering::LowerRETURNADDR(SDValue Op,
+                                              SelectionDAG &DAG) const {
+  // check the depth
+  // TODO: xtensa-gcc can handle this, by navigating through the stack, we
+  // should be able to do this too
+  assert((cast<ConstantSDNode>(Op.getOperand(0))->getZExtValue() == 0) &&
```
Comment (arsenm) [inline llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:602]: Shouldn't assert on this, it's not invalid IR. Langref states it returns "zero if it cannot be identified"

File: llvm/lib/Target/Xtensa/XtensaISelLowering.cpp [hunk 4]
```diff
@@ -722,6 +742,24 @@ SDValue XtensaTargetLowering::LowerSTACKRESTORE(SDValue Op,
                           Op.getOperand(1));
 }
 
+SDValue XtensaTargetLowering::LowerFRAMEADDR(SDValue Op,
+                                             SelectionDAG &DAG) const {
+  // check the depth
+  assert((cast<ConstantSDNode>(Op.getOperand(0))->getZExtValue() == 0) &&
```
Comment (arsenm) [inline llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:748]: Same

File: llvm/test/CodeGen/Xtensa/frameaddr-returnaddr.ll [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  ret ptr %1
+}
+
+define ptr @test_returnaddress_0() nounwind {
+; CHECK-LABEL: test_returnaddress_0:
+; CHECK:         or a2, a0, a0
+; CHECK-NEXT:    ret
+  %1 = call ptr @llvm.returnaddress(i32 0)
+  ret ptr %1
+}
```
Comment (arsenm) [inline llvm/test/CodeGen/Xtensa/frameaddr-returnaddr.ll:22]: Test the non-0 cases

File: llvm/test/CodeGen/Xtensa/frameaddr-returnaddr.ll [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+; RUN:   | FileCheck %s
+
+declare ptr @llvm.frameaddress(i32)
+declare ptr @llvm.returnaddress(i32)
+
+define ptr @test_frameaddress_0() nounwind {
+; CHECK-LABEL: test_frameaddress_0:
+; CHECK:         or a2, a1, a1
+; CHECK-NEXT:    ret
+  %1 = call ptr @llvm.frameaddress(i32 0)
```
Comment (arsenm) [inline llvm/test/CodeGen/Xtensa/frameaddr-returnaddr.ll:12]: Use named values in tests

File: llvm/test/CodeGen/Xtensa/frameaddr-returnaddr.ll [hunk 7]
```diff
@@ -0,0 +1,22 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=xtensa -verify-machineinstrs < %s \
```
Comment (arsenm) [inline llvm/test/CodeGen/Xtensa/frameaddr-returnaddr.ll:2]: Don't need -verify-machineinstrs

File: llvm/lib/Target/Xtensa/XtensaISelLowering.cpp [hunk 8]
```diff
@@ -722,6 +742,24 @@ SDValue XtensaTargetLowering::LowerSTACKRESTORE(SDValue Op,
...
+         "Frame address can only be determined for current frame.");
+
+  MachineFunction &MF = DAG.getMachineFunction();
+  MachineFrameInfo &MFI = DAG.getMachineFunction().getFrameInfo();
+  MFI.setFrameAddressIsTaken(true);
+  EVT VT = Op.getValueType();
+  SDLoc DL(Op);
+
+  unsigned FrameRegister = Subtarget.getRegisterInfo()->getFrameRegister(MF);
```
Comment (arsenm) [inline llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:757]: Register

File: llvm/lib/Target/Xtensa/XtensaISelLowering.cpp (hunk 2 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaISelLowering.cpp]: Fixed

File: llvm/lib/Target/Xtensa/XtensaISelLowering.cpp (hunk 3 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaISelLowering.cpp]: Fixed. I hope  that I understand correctly your comment,

File: llvm/lib/Target/Xtensa/XtensaISelLowering.cpp (hunk 4 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaISelLowering.cpp]: Corrected

File: llvm/test/CodeGen/Xtensa/frameaddr-returnaddr.ll (hunk 5 above)
Comment (andreisfr) [reply llvm/test/CodeGen/Xtensa/frameaddr-returnaddr.ll:38]: Added new tests

File: llvm/test/CodeGen/Xtensa/frameaddr-returnaddr.ll (hunk 6 above)
Comment (andreisfr) [reply llvm/test/CodeGen/Xtensa/frameaddr-returnaddr.ll]: Corrected

File: llvm/lib/Target/Xtensa/XtensaISelLowering.cpp (hunk 8 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaISelLowering.cpp]: Corrected

File: llvm/lib/Target/Xtensa/XtensaISelLowering.cpp [hunk 9]
```diff
@@ -722,6 +743,28 @@ SDValue XtensaTargetLowering::LowerSTACKRESTORE(SDValue Op,
...
+  // This nodes represent llvm.frameaddress on the DAG.
+  // It takes one operand, the index of the frame address to return.
+  // An index of zero corresponds to the current function's frame address.
+  // An index of one to the parent's frame address, and so on.
+  // Depths > 0 not supported yet!
+  if (Op.getConstantOperandVal(0) > 0)
+    return SDValue();
+
+  MachineFunction &MF = DAG.getMachineFunction();
+  MachineFrameInfo &MFI = DAG.getMachineFunction().getFrameInfo();
```
Comment (arsenm) [inline llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:757]:
```suggestion
  MachineFrameInfo &MFI = MF.getFrameInfo();
```

File: llvm/lib/Target/Xtensa/XtensaISelLowering.cpp [hunk 10]
```diff
@@ -594,6 +594,27 @@ SDValue XtensaTargetLowering::LowerSELECT_CC(SDValue Op,
...
 
+SDValue XtensaTargetLowering::LowerRETURNADDR(SDValue Op,
+                                              SelectionDAG &DAG) const {
+  // This nodes represent llvm.returnaddress on the DAG.
+  // It takes one operand, the index of the return address to return.
+  // An index of zero corresponds to the current function's return address.
+  // An index of one to the parent's return address, and so on.
+  // Depths > 0 not supported yet!
+  if (Op.getConstantOperandVal(0) > 0)
+    return SDValue();
```
Comment (arsenm) [inline llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:605]: != 0

File: llvm/lib/Target/Xtensa/XtensaISelLowering.cpp [hunk 11]
```diff
@@ -722,6 +743,28 @@ SDValue XtensaTargetLowering::LowerSTACKRESTORE(SDValue Op,
...
 }
 
+SDValue XtensaTargetLowering::LowerFRAMEADDR(SDValue Op,
+                                             SelectionDAG &DAG) const {
+  // This nodes represent llvm.frameaddress on the DAG.
+  // It takes one operand, the index of the frame address to return.
+  // An index of zero corresponds to the current function's frame address.
+  // An index of one to the parent's frame address, and so on.
+  // Depths > 0 not supported yet!
+  if (Op.getConstantOperandVal(0) > 0)
```
Comment (arsenm) [inline llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:753]: != 0

**Review by arsenm — APPROVED** [review]

File: llvm/lib/Target/Xtensa/XtensaISelLowering.cpp (hunk 9 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaISelLowering.cpp]: Fixed

File: llvm/lib/Target/Xtensa/XtensaISelLowering.cpp (hunk 10 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:605]: Fixed

**Review by arsenm — APPROVED** [review]

**Pull request merged**